### PR TITLE
Implement distributed gradient desc scheduler

### PR DESCRIPTION
**Pull request by limetreeestate**
Signed-off-by: Suchitha Dehigaspitiya <suchithald@gmail.com>

### CODE REVIEW

File: src/ml/trainer/JasmineGraphTrainingSchedular.cpp [hunk 1]
```diff
@@ -177,3 +178,215 @@ JasmineGraphTrainingSchedular::packPartitionsToMemory(vector<pair<int, int>> par
     trainScheduler_logger.log("Packing partitions to fit memory completed", "info");
     return partitionToIteration;
 }
+
+/** Method to initiate the creation of training schedule
+ *
+ *  @param graphID ID of graph to be trained
```
Comment (miyurud) [inline src/ml/trainer/JasmineGraphTrainingSchedular.cpp:184]: Lets mention the return value as well.

File: src/ml/trainer/JasmineGraphTrainingSchedular.cpp [hunk 2]
```diff
@@ -177,3 +178,215 @@ JasmineGraphTrainingSchedular::packPartitionsToMemory(vector<pair<int, int>> par
...
+        map<int, map<int, int>> scheduledPartitionSets = bestFitPartition(partitionMemoryList, partitionWorkerMap, availableMemory);
+        scheduleForEachHost.insert(make_pair(j->second, scheduledPartitionSets));
+    }
+    return scheduleForEachHost;
+}
+
+/** Estimate memory of graph partitions using the partition details
+ *
+ * @param partitionDetails a vector of vectors containing partition ID, node count, edge count, feature count per partition
+ * @param availableMemory total memory of host
```
Comment (miyurud) [inline src/ml/trainer/JasmineGraphTrainingSchedular.cpp:285]: Lets specify the return values here

File: src/ml/trainer/JasmineGraphTrainingSchedular.cpp [hunk 3]
```diff
@@ -177,3 +178,215 @@ JasmineGraphTrainingSchedular::packPartitionsToMemory(vector<pair<int, int>> par
...
+    vector<pair<int, int>> partitionMemoryList; //Vector of estimated sizes of partitions
+
+    //Iterate through partitions and estimate sizes
+    for (vector<vector<int>>::iterator i = partitionDetails.begin(); i != partitionDetails.end(); i++) {
+        int partitionID = (*i)[0];
+        int nodeCount = (*i)[1];
+        int edgeCount = (*i)[2];
+        int featureCount = (*i)[3];
+
+        int graphSize = 8*nodeCount*edgeCount*featureCount / (1024 * 1024); //Size in hard disk
```
Comment (miyurud) [inline src/ml/trainer/JasmineGraphTrainingSchedular.cpp:300]: What does `8` here stands for?

File: src/ml/trainer/JasmineGraphTrainingSchedular.cpp [hunk 4]
```diff
@@ -177,3 +178,215 @@ JasmineGraphTrainingSchedular::packPartitionsToMemory(vector<pair<int, int>> par
...
+
+        int sizeInMem = 36 * graphSize + 2; //Size when loaded into memory
```
Comment (miyurud) [inline src/ml/trainer/JasmineGraphTrainingSchedular.cpp:302]: Why we have `36` here? How did we choose that value?

File: src/ml/trainer/JasmineGraphTrainingSchedular.cpp [hunk 5]
```diff
@@ -177,3 +178,215 @@ JasmineGraphTrainingSchedular::packPartitionsToMemory(vector<pair<int, int>> par
...
+    }
+
+    return partitionMemoryList;
+}
+
+/** Create schedule for host by best fitting in sorted order (descending)
+ *
+ * @param partitionMemoryList Vector of pairs containing partition ID and estimated size in memory
+ * @param partitionWorkerMap Map of partition id and the worker assigned to partition
+ * @param capacity total memory of host
```
Comment (miyurud) [inline src/ml/trainer/JasmineGraphTrainingSchedular.cpp:318]: Lets specify the return value here

File: src/ml/trainer/JasmineGraphTrainingSchedular.cpp [hunk 6]
```diff
@@ -177,3 +178,215 @@ JasmineGraphTrainingSchedular::packPartitionsToMemory(vector<pair<int, int>> par
...
+    while (schedule.size() < partitionMemoryList.size()) {
+        int proceedToNextIter = 0;
+        //Iterate through sorted partition list and assign to memory (best fit)
+        for (vector<pair<int, int>>::iterator i = partitionMemoryList.begin(); i != partitionMemoryList.end(); i++) {
+            int partition = i->first;
+            int mem = i->second;
+            int worker = partitionWorkerMap[partition];
+
+            //Check that partition is not already scheduled
+            if (schedule.find(partition) == schedule.end()) continue;
```
Comment (miyurud) [inline src/ml/trainer/JasmineGraphTrainingSchedular.cpp:347]: Shouldn't this be `if (schedule.find(partition) != schedule.end()) continue;` instead? According to the [documentation](http://www.cplusplus.com/reference/map/map/find/) `find()` returns an iterator if it finds the key in the map if not returns the `end()`. Therefore, if the partition is already scheduled `find()` will not return `end()`. Hence we should check that condition and continue to the top of the loop.

File: src/ml/trainer/JasmineGraphTrainingSchedular.cpp [hunk 7]
```diff
@@ -177,3 +178,215 @@ JasmineGraphTrainingSchedular::packPartitionsToMemory(vector<pair<int, int>> par
...
+}
+
+/** Create schedule for host by best fitting in sorted order (descending)
+ *
+ * @param partitionMemoryList Vector of pairs containing partition ID and estimated size in memory
+ * @param partitionWorkerMap Map of partition id and the worker assigned to partition
+ * @param capacity total memory of host
+ */
+std::map<int, map<int, int>>
+JasmineGraphTrainingSchedular::bestFitPartition(vector<pair<int, int>> partitionMemoryList, map<int, int> partitionWorkerMap, int capacity) {
```
Comment (miyurud) [inline src/ml/trainer/JasmineGraphTrainingSchedular.cpp:321]: Can we have better descriptive function name than `bestFitPartition` ?

**Review by miyurud — CHANGES_REQUESTED** [review]
I have put some comments. Please check.

File: src/ml/trainer/JasmineGraphTrainingSchedular.cpp (hunk 1 above)
Comment (limetreeestate) [reply src/ml/trainer/JasmineGraphTrainingSchedular.cpp:184]: Return value mentioned

File: src/ml/trainer/JasmineGraphTrainingSchedular.cpp (hunk 2 above)
Comment (limetreeestate) [reply src/ml/trainer/JasmineGraphTrainingSchedular.cpp:285]: Return value specified

File: src/ml/trainer/JasmineGraphTrainingSchedular.cpp (hunk 3 above)
Comment (limetreeestate) [reply src/ml/trainer/JasmineGraphTrainingSchedular.cpp:300]: Stood for the size in bits of int8 NumPy data type used when processing attribute values by default. Changed implementation to read attribute data type from metadb and change size in bits accordingly

File: src/ml/trainer/JasmineGraphTrainingSchedular.cpp (hunk 4 above)
Comment (limetreeestate) [reply src/ml/trainer/JasmineGraphTrainingSchedular.cpp:302]: Results from experiments to map the relationship between HDD size and size in memory. However, there is a typo here and the value was changed to 3.6

File: src/ml/trainer/JasmineGraphTrainingSchedular.cpp (hunk 5 above)
Comment (limetreeestate) [reply src/ml/trainer/JasmineGraphTrainingSchedular.cpp:318]: Return value specified

File: src/ml/trainer/JasmineGraphTrainingSchedular.cpp (hunk 6 above)
Comment (limetreeestate) [reply src/ml/trainer/JasmineGraphTrainingSchedular.cpp:347]: Done the changes as suggested and also included some additional condition checks required for this purpose.

File: src/ml/trainer/JasmineGraphTrainingSchedular.cpp (hunk 7 above)
Comment (limetreeestate) [reply src/ml/trainer/JasmineGraphTrainingSchedular.cpp:321]: Method name changed to `schedulePartitionsBestFit`

File: src/ml/trainer/JasmineGraphTrainingSchedular.cpp (hunk 3 above)
Comment (limetreeestate) [reply src/ml/trainer/JasmineGraphTrainingSchedular.cpp:300]: New code added to `JasmineGraphFrontEnd` and `MetisPartitioner` to accept or infer the graph node attribute data type and store in metadb

File: src/ml/trainer/JasmineGraphTrainingSchedular.cpp (hunk 1 above)
Comment (miyurud) [reply src/ml/trainer/JasmineGraphTrainingSchedular.cpp:187]: Done

File: src/ml/trainer/JasmineGraphTrainingSchedular.cpp (hunk 2 above)
Comment (miyurud) [reply src/ml/trainer/JasmineGraphTrainingSchedular.cpp]: Done

File: src/ml/trainer/JasmineGraphTrainingSchedular.cpp (hunk 3 above)
Comment (miyurud) [reply src/ml/trainer/JasmineGraphTrainingSchedular.cpp]: Ok, we should also commit the updated metadb to JasmineGraph repository. I am referring this [file](https://github.com/miyurud/jasminegraph/blob/master/metadb/jasminegraph_meta.db).

File: src/ml/trainer/JasmineGraphTrainingSchedular.cpp (hunk 4 above)
Comment (miyurud) [reply src/ml/trainer/JasmineGraphTrainingSchedular.cpp]: Can you add one comment line describing how 3.6 was derived?

File: src/ml/trainer/JasmineGraphTrainingSchedular.cpp (hunk 5 above)
Comment (miyurud) [reply src/ml/trainer/JasmineGraphTrainingSchedular.cpp:339]: Done

File: src/ml/trainer/JasmineGraphTrainingSchedular.cpp (hunk 6 above)
Comment (miyurud) [reply src/ml/trainer/JasmineGraphTrainingSchedular.cpp]: Done

File: src/ml/trainer/JasmineGraphTrainingSchedular.cpp (hunk 7 above)
Comment (miyurud) [reply src/ml/trainer/JasmineGraphTrainingSchedular.cpp]: Done

File: src/ml/trainer/JasmineGraphTrainingSchedular.cpp (hunk 3 above)
Comment (limetreeestate) [reply src/ml/trainer/JasmineGraphTrainingSchedular.cpp]: Added the new feature_type column to metadb in the most recent commit

File: src/ml/trainer/JasmineGraphTrainingSchedular.cpp (hunk 4 above)
Comment (limetreeestate) [reply src/ml/trainer/JasmineGraphTrainingSchedular.cpp]: Comment added

Comment (miyurud) [reply src/ml/trainer/JasmineGraphTrainingSchedular.cpp]: Ok

File: src/ml/trainer/JasmineGraphTrainingSchedular.cpp (hunk 3 above)
Comment (miyurud) [reply src/ml/trainer/JasmineGraphTrainingSchedular.cpp]: Ok, Done

**Review by miyurud — APPROVED** [review]

**Pull request merged**